src/raftstore/store/store.rs

```diff
@@ -672,31 +672,18 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         // we may encounter a message with larger peer id, which means
         // current peer is stale, then we should remove current peer
         let mut has_peer = false;
-        let mut async_remove = false;
-        let mut stale_peer = None;
-        let mut initialized = false;
+        let mut state = None;
         if let Some(p) = self.region_peers.get_mut(&region_id) {
             has_peer = true;
             let target_peer_id = target.get_id();
             if p.peer_id() < target_peer_id {
-                initialized = p.get_store().is_initialized();
-                async_remove = initialized;
-                if p.is_applying_snapshot() {
-                    if !p.mut_store().cancel_applying_snap() {
-                        info!(
-                            "[region {}] Stale peer {} is applying snapshot, will destroy next \
-                             time.",
-                            region_id,
-                            p.peer_id()
-                        );
-                        self.raft_metrics.message_dropped.stale_peer += 1;
+                state = match p.maybe_destroy() {
+                    None => {
+                        self.raft_metrics.message_dropped.applying_snap += 1;
                         return Ok(false);
                     }
-                    // There is no tasks in apply worker.
-                    async_remove = false;
-                }
-                p.pending_remove = true;
-                stale_peer = Some(p.peer.clone());
+                    Some(job) => Some((job, p.peer.clone())),
```

Comment:
Why not pass a peer reference?

Reply:
When job is executed, `self` should not be borrowed.